### PR TITLE
[AIR][Telemetry] Environment variable usage

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR adds telemetry for **environment variable usage** by adding 1 usage tag:

- `AIR_ENV_VARS`: JSON string representation of a list of string environment variable keys.

This will help us "deprecate" / remove environment variables, since there is currently no way of doing phasing these out, so environment variables keep accumulating. Also, if certain environment variables are always set or never set, we can consider changing the default behavior within the library.

### Comments / Questions

1. `TUNE_ORIG_WORKING_DIR` is a bad pattern that we shouldn't do again. It is impossible to check the usage of this thing, because it's not user-provided -- so collecting "how many people access this environment variable" is impossible.
2. **Question: Where can we add a note about adding a key for any new environment variables that get introduced?**

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/_internal/usage.py [hunk 1]
```diff
@@ -56,6 +58,45 @@
     "ResourceChangingScheduler",
 }
 
+AIR_ENV_VARS = {
```
Comment (matthewdeng) [inline python/ray/air/_internal/usage.py:61]: We can also add [these Train env vars](https://github.com/ray-project/ray/blob/56418911557b47021a3e5af366f7d039af5c4a03/python/ray/train/constants.py#L52-L67)

File: python/ray/air/_internal/usage.py [hunk 2]
```diff
@@ -115,3 +156,25 @@ def tag_scheduler(scheduler: "TrialScheduler"):
...
+
+    Returns:
+        bool: True if at least one environment var is supplied by the user.
+    """
+    all_env_vars = sorted(set().union(AIR_ENV_VARS, TUNE_ENV_VARS))
+
+    user_supplied_env_vars = []
+
+    for env_var in all_env_vars:
+        if os.environ.get(env_var):
```
Comment (matthewdeng) [inline python/ray/air/_internal/usage.py:172]: Do you know if there's any correctness benefit with doing this over just checking for existence?
```suggestion
        if env_var in os.environ:
```

File: python/ray/tune/tune.py [hunk 3]
```diff
@@ -555,6 +556,7 @@ class and registered trainables.
         )
 
     ray._private.usage.usage_lib.record_library_usage("tune")
+    air_usage.tag_env_vars()
```
Comment (matthewdeng) [inline python/ray/tune/tune.py:559]: Might be good to add a comment here or in the PR description explaining why this is placed here vs. anywhere else in the stack.

File: python/ray/air/_internal/usage.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/air/_internal/usage.py:172]: Ah, I think your proposal is better.

Technically possible to set ENV_VAR="", then my current logic will treat it as not set.

File: python/ray/air/_internal/usage.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/air/_internal/usage.py:61]: Good catch. I also found a few more that I missed: `COPY_DIRECTORY_CHECKPOINTS_INSTEAD_OF_MOVING_ENV`, `TRAIN_DISABLE_LAZY_CHECKPOINTING`

File: python/ray/air/tests/test_air_usage.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+
+@pytest.fixture
+def mock_record(monkeypatch):
+    import ray.air._internal.usage
+
+    recorded = {}
+
+    def mock_record_extra_usage_tag(key: TagKey, value: str):
+        recorded[key] = value
```
Comment (justinvyu) [inline python/ray/air/tests/test_air_usage.py:22]: If multiple records happen, just takes the last one. That should be fine.

File: python/ray/train/constants.py [hunk 5]
```diff
@@ -67,6 +67,15 @@
 TRAIN_ENABLE_WORKER_SPREAD_ENV = "TRAIN_ENABLE_WORKER_SPREAD"
 
 
+TRAIN_ENV_VARS = {
```
Comment (matthewdeng) [inline python/ray/train/constants.py:70]: Should we do the same for AIR/Tune?

File: python/ray/tune/tune.py [hunk 6]
```diff
@@ -556,6 +557,11 @@ class and registered trainables.
 
     ray._private.usage.usage_lib.record_library_usage("tune")
 
+    # Track environment variable usage here will also catch:
+    # 1.) Tuner usage
+    # 2.) Ray client usage (env variables are inherited by the Ray runtime env)
```
Comment (matthewdeng) [inline python/ray/tune/tune.py:562]: Also Trainer?

File: python/ray/train/constants.py (hunk 5 above)
Comment (justinvyu) [reply python/ray/train/constants.py:70]: Yes, good idea. Any ideas on **how to add a note about adding a key for any new environment variables that get introduced?**

Comment (matthewdeng) [reply python/ray/train/constants.py:70]: Nothing obvious comes to mind. Having this collocated with the other environment variables will be a good "best effort" starting point, so developers can follow by example.

**Review by matthewdeng — APPROVED** [review]

File: python/ray/air/_internal/usage.py [hunk 7]
```diff
@@ -115,3 +117,29 @@ def tag_scheduler(scheduler: "TrialScheduler"):
     assert isinstance(scheduler, TrialScheduler)
     scheduler_name = _find_class_name(scheduler, "ray.tune.schedulers", TUNE_SCHEDULERS)
     record_extra_usage_tag(TagKey.TUNE_SCHEDULER, scheduler_name)
+
+
+def tag_env_vars() -> bool:
```
Comment (thomasdesr) [inline python/ray/air/_internal/usage.py:122]: Can we name this after `ray_env_var`s or similar so its clearer that this is capturing only Ray's env vars?

File: python/ray/air/_internal/usage.py [hunk 8]
```diff
@@ -115,3 +117,29 @@ def tag_scheduler(scheduler: "TrialScheduler"):
...
+    from ray.tune.constants import TUNE_ENV_VARS
+    from ray.train.constants import TRAIN_ENV_VARS
+
+    all_env_vars = sorted(set().union(AIR_ENV_VARS, TUNE_ENV_VARS, TRAIN_ENV_VARS))
+
+    user_supplied_env_vars = []
+
+    for env_var in all_env_vars:
+        if env_var in os.environ:
+            user_supplied_env_vars.append(env_var)
```
Comment (thomasdesr) [inline python/ray/air/_internal/usage.py:138]: Ooc, why not also do set intersection here?

File: python/ray/air/_internal/usage.py (hunk 7 above)
Comment (justinvyu) [reply python/ray/air/_internal/usage.py:122]: Ah good point. I think `tag_ray_air_env_vars` is even better.

File: python/ray/air/_internal/usage.py (hunk 8 above)
Comment (justinvyu) [reply python/ray/air/_internal/usage.py:138]: Yeah, thats a nicer way!

Comment (justinvyu) [reply python/ray/air/_internal/usage.py:138]: Oh, actually. The user might have way more env variables, so I don't want to incur the cost of converting the env var keys to a set. I'd rather just check for existence and loop through the smaller set of AIR env variables. I'll keep the current way.

**Review by thomasdesr — APPROVED** [review]

**Review by krfricke — APPROVED** [review]

**Pull request merged**